tctl-snap/tctl-plugins/cmd/tctl-login/login.go

```diff
@@ -0,0 +1,236 @@
+// Copyright 2022 Canonical Ltd.
+
+package main
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+	"os/signal"
+	"syscall"
+
+	"github.com/canonical/charmed-temporal-image/tctl-plugins/cmd"
+	"github.com/google/uuid"
+	"github.com/skratchdot/open-golang/open"
+)
+
+const (
+	redirectURI  = "http://localhost:7070/oauth2callback"
+	authEndpoint = "https://accounts.google.com/o/oauth2/auth"
+	scope        = "openid profile email"
+)
+
+var (
+	server       *http.Server
+	shutdown     = make(chan struct{}) // Custom channel for graceful shutdown signal
+	state        = uuid.New().String()
+	clientID     string
+	clientSecret string
+	codeVerifier string
+)
+
+func main() {
+	env := os.Getenv("TCTL_ENVIRONMENT")
+	if env == "dev" {
+		fmt.Fprintf(os.Stderr, "login command not defined for development environment.\n")
+		return
+	}
+
+	var err error
+	clientID, err = cmd.ClientID()
+	if err != nil {
+		return
+	}
+
+	if clientID == "" {
+		fmt.Fprintf(os.Stderr, "no google-client-id found for %v environment. use 'sudo snap set tctl %v-google-client-id=\"<client_id>\"'.\n", env, env)
+		return
+	}
+
+	clientSecret, err = cmd.ClientSecret()
+	if err != nil {
+		return
+	}
+
+	if clientSecret == "" {
+		fmt.Fprintf(os.Stderr, "no google-client-secret found for %v environment. use 'sudo snap set tctl %v-google-client-secret=\"<client_secret>\"'.\n", env, env)
+		return
+	}
+
+	token, _ := cmd.FetchValidToken(clientID, clientSecret)
+	if token != "" {
+		os.Exit(1)
+	}
+
+	if err := getToken(); err != nil {
+		fmt.Fprintf(os.Stderr, "unable to login: %s\n", err)
+		os.Exit(1)
+	}
+}
+
+// getToken returns a valid Google OAuth 2.0 access token.
+func getToken() error {
+	authURL := generateAuthURL()
+
+	// Automatically open the browser for the user to log in
+	if err := open.Run(authURL); err != nil {
+		fmt.Println("Failed to open the browser:", err)
+		return err
+	}
+
+	shutdownChan := make(chan os.Signal, 1)
+	signal.Notify(shutdownChan, os.Interrupt, syscall.SIGINT, syscall.SIGTERM)
+
+	// Start an HTTP server to handle the callback
+	go func() {
+		http.HandleFunc("/oauth2callback", handleCallback)
+		server = &http.Server{Addr: ":7070"}
```

Comment:
what if the port is taken ?
you could do something like
func GetFreePort() (int, error) {
        addr, err := net.ResolveTCPAddr("tcp", "localhost:0")
        if err != nil {
                return 0, err
        }
 
        l, err := net.ListenTCP("tcp", addr)
        if err != nil {
                return 0, err
        }
        defer l.Close()
        return l.Addr().(*net.TCPAddr).Port, nil
}

Reply:
Good point, i've added this method. Thanks!